src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesCcm.Android.cs

```diff
@@ -0,0 +1,182 @@
+                    }
+
+                    if (!Interop.Crypto.EvpCipherFinalEx(
+                        ctx,
+                        ciphertextAndTag.Slice(ciphertextBytesWritten),
+                        out int bytesWritten))
+                    {
+                        throw new CryptographicException();
+                    }
+
+                    ciphertextBytesWritten += bytesWritten;
+
+                    // NOTE: Android appends tag to the end of the ciphertext in case of CCM/GCM and "encryption" mode
+
+                    if (ciphertextBytesWritten != ciphertextAndTag.Length)
+                    {
+                        Debug.Fail($"CCM encrypt wrote {ciphertextBytesWritten} of {ciphertextAndTag.Length} bytes.");
+                        throw new CryptographicException();
+                    }
+
+                    ciphertextAndTag[..ciphertext.Length].CopyTo(ciphertext);
+                    ciphertextAndTag[ciphertext.Length..].CopyTo(tag);
+                }
+                finally
+                {
+                    if (rented != null)
+                    {
+                        CryptoPool.Return(rented, ciphertext.Length + tag.Length);
+                    }
+                }
+            }
+        }
+
+        private void DecryptInternal(
+            ReadOnlySpan<byte> nonce,
+            ReadOnlySpan<byte> ciphertext,
+            ReadOnlySpan<byte> tag,
+            Span<byte> plaintext,
+            ReadOnlySpan<byte> associatedData)
+        {
+            using (SafeEvpCipherCtxHandle ctx = Interop.Crypto.EvpCipherCreatePartial(GetCipher(_key.Length * 8)))
+            {
+                if (ctx.IsInvalid)
+                {
+                    throw new CryptographicException();
+                }
+                Interop.Crypto.EvpCipherSetNonceLength(ctx, nonce.Length);
+
+                if (!Interop.Crypto.EvpCipherSetTagLength(ctx, tag.Length))
+                {
+                    throw new CryptographicException();
+                }
+
+                Interop.Crypto.EvpCipherSetKeyAndIV(ctx, _key, nonce, Interop.Crypto.EvpCipherDirection.Decrypt);
+
+                if (associatedData.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdateAAD(ctx, associatedData))
+                    {
+                        throw new CryptographicException();
+                    }
+                }
+
+                if (!Interop.Crypto.EvpCipherUpdate(ctx, plaintext, out int plaintextBytesWritten, ciphertext))
+                {
+                    plaintext.Clear();
+                    throw new CryptographicException(SR.Cryptography_AuthTagMismatch);
+                }
+
+                if (!Interop.Crypto.EvpCipherUpdate(ctx, plaintext.Slice(plaintextBytesWritten), out int bytesWritten, tag))
+                {
+                    plaintext.Clear();
+                    throw new CryptographicException(SR.Cryptography_AuthTagMismatch);
+                }
```

Comment:
It's weird to me that these are plaintext.Clear, then later it's ZeroMemory.  Let's use ZeroMemory for both, since it's crypto code.

Reply:
It's also weird to me that these are going to report AuthTagMismatch.  The first call it can't be the auth tag, and while the second could be it doesn't feel like it's checked until CipherFinal.